File: Java/24_Graph-II/H-TUF-Floyd warshall algorithm.java
// TUF: Floyd Warshall Algorithm
// https://takeuforward.org/plus/dsa/problems/floyd-warshall-algorithm

/*
Problem Statement:
Given a weighted, directed graph with V vertices represented as an adjacency matrix.
Find the shortest distances between every pair of vertices.
The matrix[i][j] denotes the weight of the edge from i to j.
If matrix[i][j] = -1, it means there is no edge from i to j.

Approach - Floyd-Warshall Algorithm:
1. The algorithm uses Dynamic Programming approach
2. Create a distance matrix initially same as input matrix
3. Try every vertex as an intermediate vertex
4. For each pair (i, j), check if going through vertex k gives shorter path
5. Update distance[i][j] = min(distance[i][j], distance[i][k] + distance[k][j])
6. After V iterations, we get shortest distances between all pairs

Formula: dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]) for all k

Time Complexity: O(V³) - three nested loops
Space Complexity: O(1) - in-place modification (or O(V²) if creating new matrix)

Key Points:
- Works with negative weights
- Can detect negative cycles
- Finds shortest path between all pairs
*/

class Solution {
    public void shortestDistance(int[][] matrix) {
        int n = matrix.length;
        
        // Step 1: Replace -1 with a large value (infinity) 
        // except for diagonal elements
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == -1) {
                    matrix[i][j] = (int) 1e9;
                }
                if (i == j) {
                    matrix[i][j] = 0;
                }
            }
        }
        
        // Step 2: Floyd-Warshall Algorithm
        // Try every vertex as intermediate vertex
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    // If vertex k is on the shortest path from i to j
                    // then update the value of matrix[i][j]
                    matrix[i][j] = Math.min(
                        matrix[i][j], 
                        matrix[i][k] + matrix[k][j]
                    );
                }
            }
        }
        
        // Step 3: Check for negative cycle (optional)
        // If matrix[i][i] < 0 for any i, negative cycle exists
        for (int i = 0; i < n; i++) {
            if (matrix[i][i] < 0) {
                System.out.println("Negative cycle detected!");
                return;
            }
        }
        
        // Step 4: Replace infinity back with -1
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] >= (int) 1e9) {
                    matrix[i][j] = -1;
                }
            }
        }
    }

    // Helper method to print matrix
    public static void printMatrix(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                if (matrix[i][j] == -1) {
                    System.out.print("INF ");
                } else {
                    System.out.print(matrix[i][j] + " ");
                }
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Test Case 1
        System.out.println("Test Case 1:");
        int[][] matrix1 = {
            {0, 3, -1, 7},
            {8, 0, 2, -1},
            {5, -1, 0, 1},
            {2, -1, -1, 0}
        };
        
        System.out.println("Original Matrix:");
        printMatrix(matrix1);
        
        sol.shortestDistance(matrix1);
        
        System.out.println("\nShortest Distance Matrix:");
        printMatrix(matrix1);
        
        // Test Case 2
        System.out.println("\n\nTest Case 2:");
        int[][] matrix2 = {
            {0, 5, -1, 10},
            {-1, 0, 3, -1},
            {-1, -1, 0, 1},
            {-1, -1, -1, 0}
        };
        
        System.out.println("Original Matrix:");
        printMatrix(matrix2);
        
        sol.shortestDistance(matrix2);
        
        System.out.println("\nShortest Distance Matrix:");
        printMatrix(matrix2);
        
        // Test Case 3 - Simple example
        System.out.println("\n\nTest Case 3:");
        int[][] matrix3 = {
            {0, 2, -1, -1},
            {1, 0, 3, -1},
            {-1, -1, 0, -1},
            {3, 5, 4, 0}
        };
        
        System.out.println("Original Matrix:");
        printMatrix(matrix3);
        
        sol.shortestDistance(matrix3);
        
        System.out.println("\nShortest Distance Matrix:");
        printMatrix(matrix3);
    }
}

File: JavaScript/24_Graph-II/H-TUF-Floyd warshall algorithm.js
/**
 * TUF: Floyd-Warshall Algorithm
 * https://takeuforward.org/plus/dsa/problems/floyd-warshall
 *
 * All-pairs shortest path algorithm.
 * Handles negative edges (but not negative cycles).
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(V³) – Three nested loops
 * Space Complexity | O(V²) – Distance matrix (can be done in-place)
 */

/**
 * Floyd-Warshall Algorithm for all-pairs shortest paths
 * @param {number[][]} dist - Distance matrix (modify in-place)
 *        dist[i][j] = weight of edge i->j or Infinity if no edge
 *        dist[i][i] = 0
 */
function floydWarshall(dist) {
    const V = dist.length;
    const INF = 1e8;
    
    // Replace -1 with Infinity for no edge (common input format)
    for (let i = 0; i < V; i++) {
        for (let j = 0; j < V; j++) {
            if (dist[i][j] === -1 && i !== j) {
                dist[i][j] = INF;
            }
        }
    }
    
    // Main algorithm: try each vertex as intermediate
    for (let k = 0; k < V; k++) {
        for (let i = 0; i < V; i++) {
            for (let j = 0; j < V; j++) {
                if (dist[i][k] < INF && dist[k][j] < INF) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }
    
    // Replace Infinity back to -1 (if needed for output format)
    for (let i = 0; i < V; i++) {
        for (let j = 0; j < V; j++) {
            if (dist[i][j] >= INF) {
                dist[i][j] = -1;
            }
        }
    }
}

/**
 * Floyd-Warshall with path reconstruction
 */
function floydWarshallWithPath(matrix) {
    const V = matrix.length;
    const INF = 1e8;
    const dist = matrix.map(row => [...row]);
    const next = Array.from({ length: V }, () => new Array(V).fill(-1));
    
    // Initialize next matrix
    for (let i = 0; i < V; i++) {
        for (let j = 0; j < V; j++) {
            if (dist[i][j] !== -1 && dist[i][j] !== 0) {
                next[i][j] = j;
            }
            if (dist[i][j] === -1) {
                dist[i][j] = INF;
            }
        }
    }
    
    // Main algorithm
    for (let k = 0; k < V; k++) {
        for (let i = 0; i < V; i++) {
            for (let j = 0; j < V; j++) {
                if (dist[i][k] < INF && dist[k][j] < INF && 
                    dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                    next[i][j] = next[i][k];
                }
            }
        }
    }
    
    // Reconstruct path from u to v
    function getPath(u, v) {
        if (next[u][v] === -1) return [];
        const path = [u];
        while (u !== v) {
            u = next[u][v];
            path.push(u);
        }
        return path;
    }
    
    // Replace Infinity with -1
    for (let i = 0; i < V; i++) {
        for (let j = 0; j < V; j++) {
            if (dist[i][j] >= INF) dist[i][j] = -1;
        }
    }
    
    return { dist, getPath };
}

/**
 * Detect negative cycle using Floyd-Warshall
 */
function hasNegativeCycle(matrix) {
    const V = matrix.length;
    const INF = 1e8;
    const dist = matrix.map(row => row.map(x => x === -1 ? INF : x));
    
    for (let k = 0; k < V; k++) {
        for (let i = 0; i < V; i++) {
            for (let j = 0; j < V; j++) {
                if (dist[i][k] < INF && dist[k][j] < INF) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }
    
    // Negative cycle exists if diagonal has negative value
    for (let i = 0; i < V; i++) {
        if (dist[i][i] < 0) return true;
    }
    return false;
}

// Test cases
// 4 vertices, -1 means no direct edge
const matrix = [
    [0, 3, -1, 5],
    [2, 0, -1, 4],
    [-1, 1, 0, -1],
    [-1, -1, 2, 0]
];

floydWarshall(matrix);
console.log("All-pairs shortest paths:");
matrix.forEach(row => console.log(row));

// With path reconstruction
const matrix2 = [
    [0, 3, -1, 5],
    [2, 0, -1, 4],
    [-1, 1, 0, -1],
    [-1, -1, 2, 0]
];
const result = floydWarshallWithPath(matrix2);
console.log("\nPath from 0 to 2:", result.getPath(0, 2));
console.log("Path from 3 to 1:", result.getPath(3, 1));